src/pages/_document.tsx

```diff
@@ -1,6 +1,36 @@
-import Document, { Html, Head, Main, NextScript } from 'next/document'
+import type { DocumentContext, DocumentInitialProps } from 'next/document'
+import NextDocument, { Html, Head, Main, NextScript } from 'next/document'
+import { ServerStyleSheet } from 'styled-components'
+
+class Document extends NextDocument {
+  static async getInitialProps(
+    ctx: DocumentContext
+  ): Promise<DocumentInitialProps> {
+    const sheet = new ServerStyleSheet()
+    const originalRenderPage = ctx.renderPage
+
+    try {
+      ctx.renderPage = () =>
+        originalRenderPage({
+          enhanceApp: (App) => (props) =>
+            sheet.collectStyles(<App {...props} />),
```

Comment:
`enhanceApp` is a higher-order component API of Next.js that allows us to intercept `App` rendering entirely. In this case, we basically:

1. Render the `App` providing it the same `props`
2. Wrap the result in this `sheet.collectStyles`. This function will:
   1. Receives an element tree (created by executing `App`)
   2. Walk all the way down, identify styled components (those created with `styled.div\`\``, etc)
   3. Get the generated CSS of each found styled component
   4. Save into "sheet" all CSS found for later usage